pkg/nodeidentity/hetzner/identify.go

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hetzner
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/hetznercloud/hcloud-go/hcloud"
+	corev1 "k8s.io/api/core/v1"
+	expirationcache "k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"k8s.io/kops/pkg/apis/kops"
+	"k8s.io/kops/pkg/nodeidentity"
+	"k8s.io/kops/pkg/nodelabels"
+	"k8s.io/kops/upup/pkg/fi/cloudup/hetzner"
+)
+
+const (
+	cacheTTL = 60 * time.Minute
+)
+
+// nodeIdentifier identifies a node from Hetzner Cloud
+type nodeIdentifier struct {
+	client       *hcloud.Client
+	cache        expirationcache.Store
+	cacheEnabled bool
+}
+
+// New creates and returns a nodeidentity.Identifier for Nodes running on Hetzner Cloud
+func New(CacheNodeidentityInfo bool) (nodeidentity.Identifier, error) {
+	hcloudToken := os.Getenv("HCLOUD_TOKEN")
+	if hcloudToken == "" {
+		return nil, fmt.Errorf("%s is required", "HCLOUD_TOKEN")
+	}
+	opts := []hcloud.ClientOption{
+		hcloud.WithToken(hcloudToken),
+	}
+	hcloudClient := hcloud.NewClient(opts...)
+
+	return &nodeIdentifier{
+		client:       hcloudClient,
+		cache:        expirationcache.NewTTLStore(stringKeyFunc, cacheTTL),
+		cacheEnabled: CacheNodeidentityInfo,
+	}, nil
+}
+
+// IdentifyNode queries Hetzner Cloud for the node identity information
+func (i *nodeIdentifier) IdentifyNode(ctx context.Context, node *corev1.Node) (*nodeidentity.Info, error) {
+	providerID := node.Spec.ProviderID
+	if providerID == "" {
+		return nil, fmt.Errorf("providerID not set for node %s", node.Name)
+	}
+	if !strings.HasPrefix(providerID, "hcloud://") {
+		return nil, fmt.Errorf("providerID %q not recognized for node %s", providerID, node.Name)
+	}
+
+	serverID := strings.TrimPrefix(providerID, "hcloud://")
+
+	// If caching is enabled try pulling nodeidentity.Info from cache before doing a Hetzner Cloud API call.
+	if i.cacheEnabled {
+		obj, exists, err := i.cache.GetByKey(serverID)
```

Comment:
Is this a unique identifier - it looks like it is a server number, rather than e.g. the name?  If so, great!